### PR TITLE
[NEW] [Device Management] Manage Devices Admin Table

### PR DESCRIPTION
**Pull request by yash-rajpal**

### CODE REVIEW

File: apps/meteor/ee/client/views/admin/deviceManagement/DevicesTable.tsx [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+					{phase === AsyncStatePhase.LOADING && <GenericTableLoadingTable headerCells={6} />}
+					{phase === AsyncStatePhase.RESOLVED &&
+						data?.sessions &&
+						data.sessions.map((session) => (
+							<DevicesRow
+								key={session._id}
+								_id={session._id}
+								username={session._user?.username || ''}
+								ip={session.ip}
+								deviceName={session?.device?.name || ''}
```
Comment (dougfabris) [inline apps/meteor/ee/client/views/admin/deviceManagement/DevicesTable.tsx:108]: we should check which values should be required and try to not just pass empty strings for all these.

File: apps/meteor/ee/client/startup/deviceManagement.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	component: lazy(() => import('../views/admin/deviceManagement/DeviceManagementRoute')),
+	ready: false,
+});
+
+onToggledFeature('device-management', {
+	up: () =>
+		Meteor.startup(() => {
+			registerAdminSidebarItem({
+				href: '/admin/device-management',
+				i18nLabel: 'Device Management',
```
Comment (dougfabris) [inline apps/meteor/ee/client/startup/deviceManagement.ts:19]:
```suggestion
				i18nLabel: 'Device_Management',
```

File: apps/meteor/ee/client/startup/deviceManagement.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+				icon: 'mobile',
+				permissionGranted: () => hasAllPermission('view-device-management'),
+			});
+			registerRoute();
+		}),
+	down: () =>
+		Meteor.startup(() => {
+			unregisterAdminSidebarItem('Device Management');
```
Comment (dougfabris) [inline apps/meteor/ee/client/startup/deviceManagement.ts:27]:
```suggestion
			unregisterAdminSidebarItem('Device_Management');
```

File: apps/meteor/ee/client/views/admin/deviceManagement/DeviceInfoContextualBar.tsx [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	onReload,
+}: DeviceInfoContextualBarProps): ReactElement => {
+	const t = useTranslation();
+	const deviceManagementRouter = useRoute('device-management');
+	const formatDateAndTime = useFormatDateAndTime();
+	const dispatchToastMessage = useToastMessageDispatch();
+	const setModal = useSetModal();
+
+	const { name: clientName, os } = device || {};
+	const { username, name } = _user;
```
Comment (dougfabris) [inline apps/meteor/ee/client/views/admin/deviceManagement/DeviceInfoContextualBar.tsx:34]: if there's no user, it's going to break the page, we should prevent it even in the render

File: apps/meteor/ee/client/views/admin/deviceManagement/DeviceInfoContextualBar.tsx [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+	const userPresence = usePresence(userId);
+
+	const closeContextualBar = useCallback((): void => deviceManagementRouter.push({}), [deviceManagementRouter]);
```
Comment (dougfabris) [inline apps/meteor/ee/client/views/admin/deviceManagement/DeviceInfoContextualBar.tsx:37]:
```suggestion
	const handleCloseContextualBar = useCallback((): void => deviceManagementRouter.push({}), [deviceManagementRouter]);
```

File: apps/meteor/ee/client/views/admin/deviceManagement/DevicesRow.tsx [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+				<DeviceIcon deviceType={deviceType} />
+				{deviceName}
+			</TableCell>
+			<TableCell>{`${deviceOSName} ${deviceOSVersion}`}</TableCell>
+			<TableCell withTruncatedText>{username}</TableCell>
+			{mediaQuery && <TableCell>{formatDateAndTime(loginAt)}</TableCell>}
+			{mediaQuery && <TableCell withTruncatedText>{_id}</TableCell>}
+			{mediaQuery && <TableCell>{ip}</TableCell>}
+			<TableCell onClick={(e): void => e.stopPropagation()}>
+				<Menu
```
Comment (dougfabris) [inline apps/meteor/ee/client/views/admin/deviceManagement/DevicesRow.tsx:108]: Would be nice to have, rendering the `Menu` only if there's more than one action if just one renders just the button!

File: apps/meteor/packages/rocketchat-i18n/i18n/en.i18n.json [hunk 7]
```diff
@@ -918,6 +920,8 @@
   "Click_to_load": "Click to load",
   "Client_ID": "Client ID",
   "Client_Secret": "Client Secret",
+  "Client": "Client",
+  "Clients": "Clients",
```
Comment (dougfabris) [inline apps/meteor/packages/rocketchat-i18n/i18n/en.i18n.json:924]:
```suggestion

```

**Review by dougfabris — CHANGES_REQUESTED** [review]
Awesome job here, congrats!
Some points to take a look at:
1. There's an overall issue related to TOTP when listing (`sessions/list.all`) and when getting info about the device (`sessions/info.admin`) it returning bad requests for some reason even when I set the TOTP auth, I would suggest taking a look on how it behaves without TOTP configured as well.

2. I'm missing the empty state when there are no devices. Additionally, I would suggest taking a look at the filter, the query is not filtering properly
Screen Shot 2022-06-20 at 21 49 56

File: apps/meteor/ee/client/views/admin/deviceManagement/DevicesTable.tsx [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+		);
+	}
+
+	return (
+		<>
+			<FilterByText onChange={({ text }): void => setText(text)} />
+			<GenericTable>
+				<GenericTableHeader>{headers}</GenericTableHeader>
+				<GenericTableBody>
+					{phase === AsyncStatePhase.LOADING && <GenericTableLoadingTable headerCells={6} />}
```
[the remark this answers is not in the document]
Comment (dougfabris) [reply apps/meteor/ee/client/views/admin/deviceManagement/DevicesTable.tsx:99]: *product team decided to move with generic one*

File: apps/meteor/ee/client/views/admin/deviceManagement/DeviceIcon.tsx [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import { Box, Icon } from '@rocket.chat/fuselage';
+import React, { ComponentProps, ReactElement } from 'react';
+
+const iconMap: Record<string, ComponentProps<typeof Icon>['name']> = {
+	browser: 'desktop',
+	mobile: 'mobile',
+};
+
+const DeviceIcon = ({ deviceType }: { deviceType: string }): ReactElement => (
```
[the remark this answers is not in the document]
Comment (yash-rajpal) [reply apps/meteor/ee/client/views/admin/deviceManagement/DeviceIcon.tsx:9]: I agree with you, but I checked with the backend friends, they need this type to be string, its scope is bigger than MDM, and we extract the deviceType from userAgent so its difficult to restrict this type.

[the remark this answers is not in the document]
Comment (dougfabris) [reply apps/meteor/ee/client/views/admin/deviceManagement/DeviceIcon.tsx:9]: got it! 👍

File: apps/meteor/ee/client/views/admin/deviceManagement/DeviceManagementRoute.tsx [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import { usePermission } from '@rocket.chat/ui-contexts';
+import React, { ReactElement } from 'react';
+
+import NotAuthorizedPage from '../../../../../client/views/notAuthorized/NotAuthorizedPage';
+import DeviceManagementPage from './DeviceManagementPage';
+
+const DeviceManagement = (): ReactElement => {
+	const canViewDeviceManagement = usePermission('view-engagement-dashboard');
```
[the remark this answers is not in the document]
Comment (yash-rajpal) [reply apps/meteor/ee/client/views/admin/deviceManagement/DeviceManagementRoute.tsx:8]: Ouch!!

**Review by dougfabris — APPROVED** [review]

**Pull request merged**